### PR TITLE
Fixes for memory leaks and crashes

### LINKED ISSUE #3482: Workflow failed - fault_injection_full
[Failed Run](https://github.com/microsoft/ebpf-for-windows/actions/runs/8796973786)
[Codebase](https://github.com/microsoft/ebpf-for-windows/tree/c2419acd662b58a816e42e4062885834dfb576ad)
Test name - `fault_injection_full`

### PR DESCRIPTION
**Pull request by saxena-anurag**
Fixes #3482

## Description

This PR various memory leaks and crashes reported by fault injection module.

## Testing

Existing CICD

## Documentation

No

## Installation

No

### CODE REVIEW

**Review by shankarseal — DISMISSED** [review]

**Review by matthewige — APPROVED** [review]

**Review by mikeagun — DISMISSED** [review]

File: libs/shared/shared_common.c [hunk 1]
```diff
@@ -242,6 +242,10 @@ _duplicate_program_descriptor(
     program_type_descriptor_copy->header.version = EBPF_PROGRAM_TYPE_DESCRIPTOR_CURRENT_VERSION;
     program_type_descriptor_copy->header.size = EBPF_PROGRAM_TYPE_DESCRIPTOR_CURRENT_VERSION_SIZE;
 
+    // Initialize all pointers to NULL.
+    program_type_descriptor_copy->name = NULL;
+    program_type_descriptor_copy->context_descriptor = NULL;
```
Comment (dthaler) [inline libs/shared/shared_common.c:247]: Line 235 does ebpf_allocate to get the space.
ebpf_allocate is cxplat_allocate(CXPLAT_POOL_FLAG_NON_PAGED, size, EBPF_POOL_TAG_DEFAULT).
cxplat_allocate zero-initializes all memory allocated, unless the CXPLAT_POOL_FLAG_UNINITIALIZED flag is passed, which this doesn't.  So why this change?

Comment (saxena-anurag) [reply libs/shared/shared_common.c:247]: There is a `memcpy` at line 241 that copies the original struct to new struct. That also copies the embedded pointers to the new struct, and that causes double free issues.

This change is resetting all the embedded pointers to NULL before proceeding.
Same pattern at other places in this PR.

Comment (dthaler) [reply libs/shared/shared_common.c:247]: This is redundant initialization:
```
    program_type_descriptor_copy->name = NULL;
    program_type_descriptor_copy->name = cxplat_duplicate_string(program_type_descriptor->name);
```

**Review by dthaler — DISMISSED** [review]

**Review by mtfriesen — APPROVED** [review]

**Comment by saxena-anurag** [comment]
Hit fault injection failures after the last commit. Investigating.

**Review by gtrevi — APPROVED** [review]

**Review by mikeagun — APPROVED** [review]

**Pull request merged**